### PR TITLE
Implement UTXO set data structure

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Things to consider:

~- The storage is split in two, the hot cache (hashmap) and the cold storage (SQLite). Moving from the hot cache to the cold storage is done by maintaining an array acting as a queue. But this is very basic, and we could go with a linked-list instead.~
~- On shutdown, I just save everything into the cold store (the SQLite database). But ideally we should be able to restore the hot items (might not be super-important on node startup, we'll see..).~

Edit: I've removed the "hot cache", it was premature optimization. And it actually hid a bug in SQLite indexing.

- If the UTXO database couldn't be loaded from disk, we should attempt to either retrieve it from other nodes (introduce getUtxoSet() entry-point), or rebuild it from the blockchain data we have (much more costly though).

### CODE REVIEW

**Comment by Geod24** [comment]
First style nitpick: `Utxo` => `UTXO`

**Comment by AndrejMitrovic** [comment]
Rebased. I'm still not super happy with it, most of my focus was on the other task (block validation).

**Comment by AndrejMitrovic** [comment]
Looks like there are issues with connections in our system integrations test:

> object.Exception@submodules/vibe-core/source/vibe/core/net.d(229): Failed to connect to 127.0.0.1:4000: refused

**Comment by AndrejMitrovic** [comment]
This is missing some tests, and needs a rebase.

**Comment by AndrejMitrovic** [comment]
This pull request will have to be split up. While adding tests, I've realized I need to be able to skip over transactions which double-spend from the pool. I have a set of working commits, I'll submit the PRs.

**Comment by AndrejMitrovic** [comment]
This PR now requires https://github.com/bpfkorea/agora/pull/274 and https://github.com/bpfkorea/agora/pull/275.

Once those two are merged, I'll rebase this one.

It now includes a double-spend networking test-case.

**Comment by AndrejMitrovic** [comment]
Ok this is unblocked and ready for review.

**Comment by AndrejMitrovic** [comment]
There is one issue with the current implementation. The block creation is triggered by an incoming Transaction (in `acceptTransaction()`). This means whenever a new transaction comes in, and the pool contains more than 8 transactions, a block will be created.

But this also means blocks will be created at the fastest rate possible, there will never be a case where the transaction pool will contain more than 8 valid non-spending transactions.

This will have to be changed so the block is created at specific intervals (perhaps time intervals). Actually, this depends a lot on the consensus (settlement layer), so I think this should be worked on later when we work on consensus, and not in this PR.

**Comment by Geod24** [comment]
Erg sorry, needs another rebase

**Comment by AndrejMitrovic** [comment]
Rebased.

Btw, I don't know if the hot/cold cache thing is worth its current complexity. We could optimize this later and just use SQLite as-is.

**Comment by AndrejMitrovic** [comment]
Update: Removed the "hot cache". It uncovered a bug in SQLite indexing (it was an off-by-one error).

**Comment by AndrejMitrovic** [comment]
Resolved conflicts again.

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: Message mentions "UtxoSet" (UTXOSet), and the "hot cache" which have been removed.

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: `ref` ?

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: You can just use `ulong(...)` at call site though, but I don't mind if we have a function for it.

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: Mismatch

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: `scope (failure) assert(0);` at the beginning of the function

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: Ditto `scope (failure)` after the `serializePart` call

File: source/agora/consensus/data/UtxoSet.d
Comment (Geod24) [inline source/agora/consensus/data/UtxoSet.d]: Ditto `scope (failure)`

File: source/agora/test/Ledger.d
Comment (Geod24) [inline source/agora/test/Ledger.d]: ?

File: source/agora/test/Ledger.d
Comment (Geod24) [inline source/agora/test/Ledger.d]: :O

File: source/agora/test/Ledger.d
Comment (Geod24) [inline source/agora/test/Ledger.d:269]: I'd add a check that the tx is valid after this

File: source/agora/test/Ledger.d
Comment (Geod24) [inline source/agora/test/Ledger.d]: !

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d]: This is quite dangerous, but okay for now. Maybe open an issue about it ?

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d:172]: Hum, should double spend end up in the pool ? We have an issue for that, maybe link it.

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d]: `scope (failure) ...`

**Review by Geod24 — COMMENTED** [review]
Nice

File: source/agora/test/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/test/Ledger.d]: https://youtu.be/Yo1Oi2bilag

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: Badly named method, I'll rename it.

File: source/agora/consensus/data/UtxoSet.d
Comment (AndrejMitrovic) [reply source/agora/consensus/data/UtxoSet.d]: For optimization? I could do `const ref` I guess.

File: source/agora/consensus/data/UtxoSet.d
Comment (AndrejMitrovic) [reply source/agora/consensus/data/UtxoSet.d]: I originally wrote it like that, but then I realized I had three call sites where I'm casting to `ulong`, and I want to avoid adding duplicate comments at the call site explaining why it's done this way.

Also, tracking down the hashing "bug" took a good portion of my time.

File: source/agora/test/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/test/Ledger.d:269]: Ah, that's a good idea!

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d:172]: Actually this is a really good point, because it opens a DoS vector.

**Comment by AndrejMitrovic** [comment]
Made some changes: https://github.com/bpfkorea/agora/compare/2c739737594bce742651b69bdb1f964d53612688..651125ca221fb8fcd1b7aadc4829fd3640217ea6

The `reset()` function was a leaky abstraction, I've removed it. :)

**Comment by AndrejMitrovic** [comment]
There is an issue with this code, it's not re-entrant.

There could be a simple way to do this, I could return a struct with a set. However, that allocates. We could perhaps use a buffer returned from some memory allocator.. Anyway for now we could go with allocations, and work on improving it later.

@Geod24 do you want me to work on this in a separate PR, or as part of this? I don't want to delay further progress on the freezing UTXO, for example.

**Comment by Geod24** [comment]
Let's leave it for a following PR.

File: source/agora/node/Ledger.d
Comment (Geod24) [reply source/agora/node/Ledger.d]: 👋

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: Love it. (the feature)

I replaced it everywhere else but here, missed it.

**Comment by AndrejMitrovic** [comment]
Rebased again.

File: source/agora/node/Ledger.d
Comment (Geod24) [reply source/agora/node/Ledger.d]: One of Kenji's many stroke of genius

**Review by Geod24 — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
Got a stack trace of where it's stuck waiting forever locally:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSEGV
  * frame #0: 0x00007fff64fed86a libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff650ac56e libsystem_pthread.dylib`_pthread_cond_wait + 722
    frame #2: 0x0000000100712c60 agora`_D4core4sync9condition9Condition4waitMFZv + 32
    frame #3: 0x000000010009f94e agora`_D3std11concurrency10MessageBox__T3getTDFNaNbNiNfS6geod249LocalRest8ResponseZvTPFNaNiNfCQDiQDh14LinkTerminatedZvTPFNaNiNfCQEqQEp15OwnerTerminatedZvTPFNfSQFv7variant__T8VariantNVmi32ZQpZvZQFxMFMQFyMQEoMQDkMQCfZb(this=0x0000000103007f20, _param_0=void delegate(geod24.LocalRest.Response _param_0) pure nothrow @nogc @safe @ 0x00007ffeefbff020, _param_1=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda2FNaNiNfCQDhQDg14LinkTerminatedZv at concurrency.d:779), _param_2=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda3FNaNiNfCQDhQDg15OwnerTerminatedZv at concurrency.d:780), _param_3=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda4FNfSQDd7variant__T8VariantNVmi32ZQpZv at concurrency.d:781)) at concurrency.d:2148:29
    frame #4: 0x000000010002577b agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZQBh at concurrency.d:775:5
    frame #5: 0x0000000100026cec agora`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl14putTransactionMFNfSQBx9consensus4data11TransactionQnZ9__lambda3MFNeZSQFbQEx8Response at LocalRest.d-mixin-706:723:33
    frame #6: 0x0000000100018ed9 agora`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl14putTransactionMFNfSQBx9consensus4data11TransactionQnZv(this=0x000000010e3689f0, _param_0=Transaction @ 0x00007ffeefbff2e0) at LocalRest.d-mixin-706:714:21
    frame #7: 0x000000010003134a agora`_D5agora4test10BanManager17__unittest_L28_C1FZ__T9__lambda5TSQCh9consensus4data11TransactionQnZQBuMFNfQBqZv(tx=Transaction @ 0x00007ffeefbff310) at BanManager.d:79:24
    frame #8: 0x0000000100030def agora`_D3std9algorithm9iteration__T4eachS5agora4test10BanManager17__unittest_L28_C1FZ9__lambda5Z__TQCmTASQCm9consensus4data11TransactionQnZQEaMFNfKQBsZEQFo8typecons__T4FlagVAyaa4_65616368ZQv(r=<unavailable>) at iteration.d:997:29
    frame #9: 0x0000000100030422 agora`_D5agora4test10BanManager17__unittest_L28_C1FZv at BanManager.d:79:5
    frame #10: 0x000000010071224b agora`_D4core7runtime18runModuleUnitTestsUZ14__foreachbody2MFPS6object10ModuleInfoZi + 27
    frame #11: 0x0000000100739f02 agora`_D2rt5minfo17moduleinfos_applyFMDFyPS6object10ModuleInfoZiZ14__foreachbody2MFKSQCz19sections_elf_shared3DSOZi + 50
    frame #12: 0x000000010073aa6a agora`_D2rt19sections_elf_shared3DSO7opApplyFMDFKSQBqQBqQyZiZi + 58
    frame #13: 0x0000000100739eac agora`_D2rt5minfo17moduleinfos_applyFMDFyPS6object10ModuleInfoZiZi + 28
    frame #14: 0x000000010072650f agora`_D6object10ModuleInfo7opApplyFMDFPSQBhQBdZiZi + 31
    frame #15: 0x0000000100712000 agora`runModuleUnitTests + 208
    frame #16: 0x000000010073153f agora`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZv + 31
    frame #17: 0x0000000100731383 agora`_d_run_main + 547
    frame #18: 0x0000000100031485 agora`main(argc=1, argv=0x00007ffeefbff9c8) at __entrypoint.d:8:21
    frame #19: 0x00007fff64eb53d5 libdyld.dylib`start + 1
    frame #20: 0x00007fff64eb53d5 libdyld.dylib`start + 1
(lldb)
```

**Comment by AndrejMitrovic** [comment]
Here's another one, slightly different:

```
rror: need to add support for DW_TAG_base_type 'char' encoded with DW_ATE = 0x10, bit_size = 8
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSEGV
  * frame #0: 0x00007fff64fed86a libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff650ac56e libsystem_pthread.dylib`_pthread_cond_wait + 722
    frame #2: 0x0000000100712c60 agora`_D4core4sync9condition9Condition4waitMFZv + 32
    frame #3: 0x000000010009f94e agora`_D3std11concurrency10MessageBox__T3getTDFNaNbNiNfS6geod249LocalRest8ResponseZvTPFNaNiNfCQDiQDh14LinkTerminatedZvTPFNaNiNfCQEqQEp15OwnerTerminatedZvTPFNfSQFv7variant__T8VariantNVmi32ZQpZvZQFxMFMQFyMQEoMQDkMQCfZb(this=0x0000000103a07f20, _param_0=void delegate(geod24.LocalRest.Response _param_0) pure nothrow @nogc @safe @ 0x00007ffeefbff010, _param_1=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda2FNaNiNfCQDhQDg14LinkTerminatedZv at concurrency.d:779), _param_2=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda3FNaNiNfCQDhQDg15OwnerTerminatedZv at concurrency.d:780), _param_3=(agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZ9__lambda4FNfSQDd7variant__T8VariantNVmi32ZQpZv at concurrency.d:781)) at concurrency.d:2148:29
    frame #4: 0x000000010002577b agora`_D3std11concurrency__T11receiveOnlyTS6geod249LocalRest8ResponseZQBqFZQBh at concurrency.d:775:5
    frame #5: 0x0000000100025c1c agora`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl8shutdownMFNfZ9__lambda2MFNeZSQDmQDi8Response at LocalRest.d-mixin-706:723:33
    frame #6: 0x0000000100025950 agora`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl8shutdownMFNfZv(this=0x000000010dc01180) at LocalRest.d-mixin-706:714:21
    frame #7: 0x000000010002a745 agora`_D5agora4test4Base14TestAPIManager8shutdownMFZ__T9__lambda1TC6geod249LocalRest__T9RemoteAPITCQDnQDkQDi7TestAPIZQBeZQCoFNfQCjZv(a=0x000000010dc01180) at Base.d:174:30
    frame #8: 0x000000010002a767 agora`_D3std9algorithm9iteration__T4eachS5agora4test4Base14TestAPIManager8shutdownMFZ9__lambda1Z__TQCmTHSQCm6common6crypto3Key9PublicKeyC6geod249LocalRest__T9RemoteAPITCQEyQEvQEt7TestAPIZQBeZQGaMFKQDqZ14__foreachbody2MFNfKQDiZi(e=<unavailable>) at iteration.d:997:29
    frame #9: 0x000000010072c163 agora`_aaApply + 83
    frame #10: 0x000000010002091a agora`_D3std9algorithm9iteration__T4eachS5agora4test4Base14TestAPIManager8shutdownMFZ9__lambda1Z__TQCmTHSQCm6common6crypto3Key9PublicKeyC6geod249LocalRest__T9RemoteAPITCQEyQEvQEt7TestAPIZQBeZQGaMFNfKQDsZEQHo8typecons__T4FlagVAyaa4_65616368ZQv(r=<unavailable>) at iteration.d:993:21
    frame #11: 0x0000000100015c88 agora`_D5agora4test4Base14TestAPIManager8shutdownMFZv(this=0x000000010dbfda60) at Base.d:174:9
    frame #12: 0x0000000100030adc agora`_D5agora4test10BanManager17__unittest_L28_C1FZv at BanManager.d:44:17
    frame #13: 0x000000010071224b agora`_D4core7runtime18runModuleUnitTestsUZ14__foreachbody2MFPS6object10ModuleInfoZi + 27
    frame #14: 0x0000000100739f02 agora`_D2rt5minfo17moduleinfos_applyFMDFyPS6object10ModuleInfoZiZ14__foreachbody2MFKSQCz19sections_elf_shared3DSOZi + 50
    frame #15: 0x000000010073aa6a agora`_D2rt19sections_elf_shared3DSO7opApplyFMDFKSQBqQBqQyZiZi + 58
    frame #16: 0x0000000100739eac agora`_D2rt5minfo17moduleinfos_applyFMDFyPS6object10ModuleInfoZiZi + 28
    frame #17: 0x000000010072650f agora`_D6object10ModuleInfo7opApplyFMDFPSQBhQBdZiZi + 31
    frame #18: 0x0000000100712000 agora`runModuleUnitTests + 208
    frame #19: 0x000000010073153f agora`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZv + 31
    frame #20: 0x0000000100731383 agora`_d_run_main + 547
    frame #21: 0x0000000100031485 agora`main(argc=1, argv=0x00007ffeefbff9c8) at __entrypoint.d:8:21
    frame #22: 0x00007fff64eb53d5 libdyld.dylib`start + 1
    frame #23: 0x00007fff64eb53d5 libdyld.dylib`start + 1
```

However both time it's in the `BanManager.d` tests. I'll try disabling that test.

**Comment by AndrejMitrovic** [comment]
@Geod24 Disabling the BanManager test gets rid of the infinite wait issue.

I think the bug might be somewhere in the handling of filtered requests in LocalRest, because the BanManager makes heavy use of that feature there. I'll submit an issue about re-enabling the BanManager tests later if you merge this PR.

**Comment by AndrejMitrovic** [comment]
I'm doing another rebuild. If this passes, I'll merge it so we can unblock everyone else that needs UTXOs.

**Comment by AndrejMitrovic** [comment]
Green again.

**Pull request merged**
